### PR TITLE
Merge 3.6.2 into 3.6 (updated cmake gcc+Asan build flags)

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Merge 3.6.2 from https://github.com/Mbed-TLS/mbedtls/pull/9694 back into `mbedtls-3.6`.

Alternative to https://github.com/Mbed-TLS/mbedtls/pull/9695.

This version has `-O2` instead of `-O3` for GCC+Asan cmake builds.

## PR checklist

- [x] **changelog** not required because: just technical git stuff
- [x] **development PR** not required because: N/A
- [x] **framework PR** not required
- [x] **3.6 PR** provided
- [x] **2.28 PR** not required because: no 2.28 release
- **tests**  provided | not required because:

### CODE REVIEW

**Review by yanesca — APPROVED** [review]
Reviewed the diff to https://github.com/Mbed-TLS/mbedtls/pull/9695 and it looks good to me.

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM, reviewed the commits from 107ea89

**Comment by gilles-peskine-arm** [comment]
@bagder I don't see a typo? We just published the advisory a few minutes ago: https://mbed-tls.readthedocs.io/en/latest/security-advisories/mbedtls-security-advisory-2024-10-1/

**Comment by bagder** [comment]
Thanks!

File: CMakeLists.txt [hunk 1]
```diff
@@ -229,7 +229,15 @@ if(CMAKE_COMPILER_IS_GNU)
...
     set(CMAKE_C_FLAGS_COVERAGE    "-O0 -g3 --coverage")
-    set(CMAKE_C_FLAGS_ASAN        "-fsanitize=address -fno-common -fsanitize=undefined -fno-sanitize-recover=all -O3")
+    # Old GCC versions hit a performance problem with test_suite_pkwrite
+    # "Private keey write check EC" tests when building with Asan+UBSan
+    # and -O3: those tests take more than 100x time than normal, with
+    # test_suite_pkwrite taking >3h on the CI. Observed with GCC 5.4 on
+    # Ubuntu 16.04 x86_64 and GCC 6.5 on Ubuntu 18.04 x86_64.
+    # GCC 7.5 and above on Ubuntu 18.04 appear fine.
+    # To avoid the performance problem, we use -O2 here. It doesn't slow
+    # down much even with modern compiler versions.
```
Comment (gilles-peskine-arm) [inline CMakeLists.txt:239]: Now filed as https://github.com/Mbed-TLS/mbedtls/issues/9699

**Pull request merged**